src/modm/ui/color/hsv.hpp

```diff
@@ -13,115 +9,119 @@
+	// TODO HueType should wrap rather than saturate
+	using HueType = GrayD<DH>;
+	using SaturationType = GrayD<DS>;
+	using T = GrayD<DV>;
 
-	constexpr HsvT(T hue, T saturation, T value) : hue(hue), saturation(saturation), value(value) {}
+	constexpr HsvD() = default;
 
-	/**
-	 * Copy Constructor 8bit->16bit
-	 */
-	template<std::unsigned_integral U>
-		requires std::is_same_v<T, uint16_t> && std::is_same_v<U, uint8_t>
-	constexpr HsvT(const HsvT<U> &hsv_other)
-		: hue(hsv_other.hue << 8), saturation(hsv_other.saturation << 8), value(hsv_other.value << 8)
+	constexpr HsvD(HueType hue, SaturationType saturation, T value)
+		: hue(hue), saturation(saturation), value(value)
 	{}
 
-	/**
-	 * Copy Constructor 16bit->8bit
-	 */
-	template<std::unsigned_integral U>
-		requires std::is_same_v<T, uint8_t> && std::is_same_v<U, uint16_t>
-	constexpr HsvT(const HsvT<U> &hsv_other)
-		: hue(hsv_other.hue >> 8), saturation(hsv_other.saturation >> 8), value(hsv_other.value >> 8)
+	template<ColorHsv C>
+	constexpr HsvD(const C &other)
+		: hue(other.hue), saturation(other.saturation), value(other.value)
 	{}
 
-	/**
-	 * Convertion Constructor for RGB Color
-	 *
-	 * @param rgb	RGB Color
-	 */
-	template<std::unsigned_integral U>
-	constexpr HsvT(const RgbT<U>& rgb);
-
-	/**
-	 * Convertion Constructor for Brightness
-	 *
-	 * @param brightness	Brightness 'Color'-object
-	 */
-	template<std::unsigned_integral U>
-	constexpr HsvT(const BrightnessT<U> gray) : hue(0), saturation(0), value(gray.value)
-	{}
+	template<ColorRgb C>
+	constexpr HsvD(const C& rgb);
+
+	template<ColorRgbStacked CS>
+ 	constexpr HsvD(const CS& rgbstacked) : HsvD(RgbD<CS::DR, CS::DG, CS::DB>(rgbstacked)) {}
+
+	void setHue(HueType hue) { this->hue = hue;}
+	void setSaturation(SaturationType saturation) { this->saturation = saturation;}
+	void setValue(T value) { this->value = value;}
 
-	/**
-	 * Convertion Constructor for RGB565 Color
-	 *
-	 * @param rgb565	RGB565 Color
-	 */
-	constexpr HsvT(const Rgb565& rgb565) : HsvT(RgbT<uint8_t>(rgb565)) {}
+	HueType getHue() const { return hue; }
+
+	SaturationType getSaturation() const { return saturation; }
+	T getValue() const { return value; }
 
 	constexpr bool
-	operator==(const HsvT<T>& other) const = default;
+	operator==(const HsvD& other) const = default;
+
+	// IMPLEMENT operator+=()
+	// IMPLEMENT operator-=()
+	// IMPLEMENT operator*=()
+
+	void invert() {
+		hue.invert();
+	}
+
+	// Output human friendly:
+	// Hue in deg, Sat in pct, Value in pct
+	void iostream_human_friendly(IOStream& io) {
+		using CalcTypeHue = modm::WideType<typename HueType::T>;
+		io << (CalcTypeHue(hue.getValue()) * 360 / hue.max) << "deg\t";
+
+		using CalcTypeSaturation = modm::WideType<typename SaturationType::T>;
+ 		io << (CalcTypeSaturation(saturation.getValue()) * 100 / saturation.max) << "%\t";
+
+		using CalcTypeValue = modm::WideType<typename T::T>;
+		io << (CalcTypeValue(value.getValue()) * 100 / value.max) << "%";
+	}
```

Comment:
Needs to be guarded by `#if __has_include(<modm/io/iostream.hpp>)` to prevent people being forced to include the expensive modm IOStream.

Reply:
I've also friended `ostream_human_friendly()` and moved it next to the regular `operator<<()` so it keeps assorted nicely.